prpl/attachments.c

```diff
@@ -192,3 +194,423 @@ void download_attachment(ChimeConnection *cxn, ChimeAttachment *att, AttachmentC
+			purple_xfer_conversation_write(xfer, _("Could not get upload url"), TRUE);
+			deep_free_upload_data(xfer);
+		}
+	} else {
+		if (!SOUP_STATUS_IS_SUCCESSFUL(msg->status_code)) {
+			const gchar *reason = msg->reason_phrase;
+
+			if (node)
+				parse_string(node, "Message", &reason);
+
+			gchar *error_msg = g_strdup_printf(_("Failed to request upload: %d %s"),
+							   msg->status_code,
+							   reason);
+			purple_xfer_conversation_write(xfer, error_msg, TRUE);
+			g_free(error_msg);
+		} else if (!node) {
+			purple_xfer_conversation_write(xfer, _("Failed to request upload"), TRUE);
+		}
+		deep_free_upload_data(xfer);
+	}
+}
+
+static void request_upload_url(ChimeConnection *self, const gchar *messaging_url, PurpleXfer *xfer)
+{
+	AttachmentUpload *data = (AttachmentUpload*)xfer->data;
+
+	JsonBuilder *jb = json_builder_new();
+	jb = json_builder_begin_object(jb);
+	jb = json_builder_set_member_name(jb, "ContentType");
+	jb = json_builder_add_string_value(jb, data->content_type);
+	jb = json_builder_end_object(jb);
+
+	SoupURI *uri = soup_uri_new_printf(messaging_url, "/uploads");
+	JsonNode *node = json_builder_get_root(jb);
+	chime_connection_queue_http_request(self, node, uri, "POST", request_upload_url_callback, xfer);
+
+	json_node_unref(node);
+	g_object_unref(jb);
+}
+
+struct FileType {
+	const gchar *file_extension;
+	const gchar *mime_type;
+};
+
+// Based on https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Complete_list_of_MIME_types
+struct FileType file_types[] = {
+	{".aac", "audio/aac"},
+	{".avi", "video/x-msvideo"},
+	{".doc", "application/msword"},
+	{".docx", "application/vnd.openxmlformats-officedocument.wordprocessingml.document"},
+	{".gif", "image/gif"},
+	{".htm", "text/html"},
+	{".html", "text/html"},
+	{".ics", "text/calendar"},
+	{".jpeg", "image/jpeg"},
+	{".jpg", "image/jpeg"},
+	{".mid", "audio/midi"},
+	{".midi", "audio/midi"},
+	{".mpeg", "video/mpeg"},
+	{".odp", "application/vnd.oasis.opendocument.presentation"},
+	{".ods", "application/vnd.oasis.opendocument.spreadsheet"},
+	{".odt", "application/vnd.oasis.opendocument.text"},
+	{".oga", "audio/ogg"},
+	{".ogv", "video/ogg"},
+	{".ogx", "application/ogg"},
+	{".png", "image/png"},
+	{".pdf", "application/pdf"},
+	{".ppt", "application/vnd.ms-powerpoint"},
+	{".pptx", "application/vnd.openxmlformats-officedocument.presentationml.presentation"},
+	{".rar", "application/x-rar-compressed"},
+	{".rtf", "application/rtf"},
+	{".svg", "image/svg+xml"},
+	{".tar", "application/x-tar"},
+	{".tif", "image/tiff"},
+	{".tiff", "image/tiff"},
+	{".wav", "audio/x-wav"},
+	{".weba", "audio/webm"},
+	{".webm", "video/webm"},
+	{".webp", "image/webp"},
+	{".xhtml", "application/xhtml+xml"},
+	{".xls", "application/vnd.ms-excel"},
+	{".xlsx", "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"},
+	{".xml", "application/xml"},
+	{".zip", "application/zip"},
+	{".7z", "application/x-7z-compressed"},
+};
```

Comment:
Must we have this for ourselves? Is there *no* library already available to us which does it?

Reply:
we could wrap the filename as a GFile, guess the content type and from it get the mime type...